k8s.go

```diff
@@ -0,0 +1,643 @@
+	act := ActionList{}
+	for key, last := range lasts {
+		if cur, ok := currs[key]; ok == false {
+			act.Del = append(act.Del, key)
+		} else if last != cur {
+			act.Upd = append(act.Upd, key)
+		} else {
+			act.Same = append(act.Same, key)
+		}
+	}
+	for key, _ := range currs {
+		if _, ok := lasts[key]; ok == false {
+			act.Add = append(act.Add, key)
+		}
+	}
+	return act
+}
+
+// populateFromKVPairs populates the vars KV map from the supplied set of
+// KVPairs.  This uses the libcalico-go compat module and serialization functions
+// to write out the KVPairs in etcdv2 format.  This works in conjunction with the
+// etcdVarClient defined below which provides a "mock" etcd backend which actually
+// just writes out data to the vars map.
+func populateFromKVPairs(kvps []*model.KVPair, vars map[string]string) {
+	// Create a etcdVarClient to write the KVP results in the vars map, using the
+	// compat adaptor to write the values in etcdv2 format.
+	client := compat.NewAdaptor(&etcdVarClient{vars: vars})
+	for _, kvp := range kvps {
+		if _, err := client.Apply(kvp); err != nil {
+			log.Error("Failed to convert k8s data to etcdv2 equivalent: %s = %s", kvp.Key, kvp.Value)
+		}
+	}
+}
+
+type IntervalProcessor struct {
+	interval int
+	k8scli   *k8sClient
+	ipam     *ipamCacheK8s
+}
+
+func (p *IntervalProcessor) IntervalLoop() error {
+	if err := p.k8scli.updatePrefix(); err != nil {
+		return err
+	}
+	if err := p.k8scli.initialNeighborConfigSetting(); err != nil {
+		return err
+	}
+	ippool, err := p.ipam.getIPPools()
+	if err != nil {
+		return err
+	}
+	p.ipam.lastIPPool = ippool
+	for {
+		log.Debug("polling")
+		p.ipam.sync()
+		p.k8scli.checkBGPConfig()
+		select {
+		case <-time.After(time.Duration(p.interval) * time.Second):
+			continue
+		}
+	}
+}
+
+type k8sClient struct {
+	node              string
+	server            *Server
+	k8scli            *kubernetes.Clientset
+	nodeBgpPeerClient resources.K8sNodeResourceClient
+	nodeBgpCfgClient  resources.K8sNodeResourceClient
+	lastBgpconfig     map[string]string
+}
+
+// create new Kubernetes client
+func NewK8sClient(s *Server) (*k8sClient, error) {
+	loadingRules := clientcmd.ClientConfigLoadingRules{}
+	config, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
+		&loadingRules, &clientcmd.ConfigOverrides{}).ClientConfig()
+	if err != nil {
+		return nil, err
+	}
+
+	// Create the clientset
+	cs, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return &k8sClient{
+		node:              os.Getenv(NODENAME),
+		server:            s,
+		k8scli:            cs,
+		nodeBgpPeerClient: resources.NewNodeBGPPeerClient(cs),
+		nodeBgpCfgClient:  resources.NewNodeBGPConfigClient(cs),
+	}, nil
+}
+
+func (c *k8sClient) updatePrefix() error {
+	var paths []*svbgptable.Path
+	node, err := c.k8scli.Nodes().Get(c.node, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+	cidr := node.Spec.PodCIDR
```

Comment:
The code assumes PodCIDR will never change. Is that ok?

Reply:
I think that it is fixed now and some kind of mechanism may be mad in the future.
At the time the codes should be changed.